<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="ie=edge">

<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="author" content="">
<meta name="description" content="Modified: July 19 2024
Errors and Exceptions 6/22/2024
There are two distinguishable types of errors: syntax errors and exceptions.
8.1 Syntax Syntax errors (or parsing errors) are errors involving the actual use of the language.
&amp;gt;&amp;gt;&amp;gt; while True print(&amp;#39;Hello World&amp;#39;) File &amp;#34;&amp;lt;stdin&amp;gt;&amp;#34;, line 1 while True print(&amp;#39;Hello world&amp;#39;) ^^^^^ SyntaxError: invalid syntax In the example above, the while statement is missing the : required to use it.
8.2 Exceptions Even if code is syntactically correct, an error may occur when attempting to execute it." />
<meta name="keywords" content="homepage, blog, computerscience, python, documentation, errors, exceptions" />
<meta name="robots" content="noodp" />
<meta name="theme-color" content="" />
<link rel="canonical" href="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-8/" />


    <title>
        
            The Python Tutorial Notes - Chapter 8 :: Mr. Pointing  — Python Teacher
        
    </title>





<link rel="stylesheet" href="/main.9a8b2bfd3b57cf24918963cf5c8025ac2284e56ea037e60d10a5981ecfcc5055.css" integrity="sha256-mosr/TtXzySRiWPPXIAlrCKE5W6gN&#43;YNEKWYHs/MUFU=">



    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
    <link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
    <link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
    <link rel="manifest" href="/site.webmanifest">
    <link rel="mask-icon" href="/safari-pinned-tab.svg" color="">
    <link rel="shortcut icon" href="/favicon.ico">
    <meta name="msapplication-TileColor" content="">



  <meta itemprop="name" content="The Python Tutorial Notes - Chapter 8">
  <meta itemprop="description" content="Modified: July 19 2024
Errors and Exceptions 6/22/2024
There are two distinguishable types of errors: syntax errors and exceptions.
8.1 Syntax Syntax errors (or parsing errors) are errors involving the actual use of the language.
&gt;&gt;&gt; while True print(&#39;Hello World&#39;) File &#34;&lt;stdin&gt;&#34;, line 1 while True print(&#39;Hello world&#39;) ^^^^^ SyntaxError: invalid syntax In the example above, the while statement is missing the : required to use it.
8.2 Exceptions Even if code is syntactically correct, an error may occur when attempting to execute it.">
  <meta itemprop="datePublished" content="2024-06-22T00:00:00+00:00">
  <meta itemprop="dateModified" content="2024-06-22T00:00:00+00:00">
  <meta itemprop="wordCount" content="1395">
  <meta itemprop="image" content="https://mrpointing.com/">
  <meta itemprop="keywords" content="Computerscience,Python,Documentation,Errors,Exceptions">

  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:image" content="https://mrpointing.com/">
  <meta name="twitter:title" content="The Python Tutorial Notes - Chapter 8">
  <meta name="twitter:description" content="Modified: July 19 2024
Errors and Exceptions 6/22/2024
There are two distinguishable types of errors: syntax errors and exceptions.
8.1 Syntax Syntax errors (or parsing errors) are errors involving the actual use of the language.
&gt;&gt;&gt; while True print(&#39;Hello World&#39;) File &#34;&lt;stdin&gt;&#34;, line 1 while True print(&#39;Hello world&#39;) ^^^^^ SyntaxError: invalid syntax In the example above, the while statement is missing the : required to use it.
8.2 Exceptions Even if code is syntactically correct, an error may occur when attempting to execute it.">



    <meta property="og:url" content="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-8/">
  <meta property="og:site_name" content="Mr. Pointing">
  <meta property="og:title" content="The Python Tutorial Notes - Chapter 8">
  <meta property="og:description" content="Modified: July 19 2024
Errors and Exceptions 6/22/2024
There are two distinguishable types of errors: syntax errors and exceptions.
8.1 Syntax Syntax errors (or parsing errors) are errors involving the actual use of the language.
&gt;&gt;&gt; while True print(&#39;Hello World&#39;) File &#34;&lt;stdin&gt;&#34;, line 1 while True print(&#39;Hello world&#39;) ^^^^^ SyntaxError: invalid syntax In the example above, the while statement is missing the : required to use it.
8.2 Exceptions Even if code is syntactically correct, an error may occur when attempting to execute it.">
  <meta property="og:locale" content="en">
  <meta property="og:type" content="article">
    <meta property="article:section" content="posts">
    <meta property="article:published_time" content="2024-06-22T00:00:00+00:00">
    <meta property="article:modified_time" content="2024-06-22T00:00:00+00:00">
    <meta property="article:tag" content="Computerscience">
    <meta property="article:tag" content="Python">
    <meta property="article:tag" content="Documentation">
    <meta property="article:tag" content="Errors">
    <meta property="article:tag" content="Exceptions">
    <meta property="og:image" content="https://mrpointing.com/">






    <meta property="article:published_time" content="2024-06-22 00:00:00 &#43;0000 UTC" />











    </head>

    
        <body>
    
    
        <div class="container">
            <header class="header">
    <span class="header__inner">
        <a href="/" style="text-decoration: none;">
    <div class="logo">
        
            <span class="logo__mark">></span>
            <span class="logo__text ">
                mr. pointing</span>
            <span class="logo__cursor" style=
                  "
                   
                   ">
            </span>
        
    </div>
</a>


        <span class="header__right">
            
                <nav class="menu">
    <ul class="menu__inner"><li><a href="/about">About</a></li><li><a href="/posts">Posts</a></li><li><a href="/projects">Projects</a></li>
    </ul>
</nav>

                <span class="menu-trigger">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
                        <path d="M0 0h24v24H0z" fill="none"/>
                        <path d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"/>
                    </svg>
                </span>
            
        </span>
    </span>
</header>


            <div class="content">
                
  <main class="post">

    <div class="post-info">
      <p>
        <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-clock">
          <circle cx="12" cy="12" r="10"></circle>
          <polyline points="12 6 12 12 16 14"></polyline>
        </svg>
        7 minutes

        
      </p>
    </div>

    <article>
      <h1 class="post-title">
        <a href="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-8/">The Python Tutorial Notes - Chapter 8</a>
      </h1>

      

      

      

      <div class="post-content">
        <p>Modified: July 19 2024</p>
<hr>
<h2 id="errors-and-exceptions"><strong>Errors and Exceptions</strong></h2>
<p>6/22/2024</p>
<p>There are two distinguishable types of errors: <em>syntax errors</em> and <em>exceptions</em>.</p>
<h3 id="81-syntax"><strong>8.1 Syntax</strong></h3>
<p>Syntax errors (or parsing errors) are errors involving the actual use of the language.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">while</span> <span style="color:#66d9ef">True</span> print(<span style="color:#e6db74">&#39;Hello World&#39;</span>)
</span></span><span style="display:flex;"><span>   File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>
</span></span><span style="display:flex;"><span>    <span style="color:#66d9ef">while</span> <span style="color:#66d9ef">True</span> print(<span style="color:#e6db74">&#39;Hello world&#39;</span>)
</span></span><span style="display:flex;"><span>               <span style="color:#f92672">^^^^^</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">SyntaxError</span>: invalid syntax
</span></span></code></pre></div><p>In the example above, the <code>while</code> statement is missing the <code>:</code> required to use it.</p>
<h3 id="82-exceptions"><strong>8.2 Exceptions</strong></h3>
<p>Even if code is syntactically correct, an error may occur when attempting to execute it. When the error is thrown, the last line of the error usually has an explanation.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#ae81ff">10</span> <span style="color:#f92672">*</span> (<span style="color:#ae81ff">1</span><span style="color:#f92672">/</span><span style="color:#ae81ff">0</span>)
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>	File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">ZeroDivisionError</span>: division by zero
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#ae81ff">4</span> <span style="color:#f92672">*</span> spam<span style="color:#f92672">*</span><span style="color:#ae81ff">3</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>	File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span>ZeroDivisonError: name <span style="color:#e6db74">&#39;spam&#39;</span> <span style="color:#f92672">is</span> <span style="color:#f92672">not</span> defined
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#e6db74">&#39;2&#39;</span> <span style="color:#f92672">*</span> <span style="color:#ae81ff">2</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>	File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">TypeError</span>: can only concantenate str(<span style="color:#f92672">not</span> <span style="color:#e6db74">&#34;int&#34;</span>) to str
</span></span></code></pre></div><p><a href="https://docs.python.org/3/library/exceptions.html#bltin-exceptions">This</a> is a document on all of the built-in exceptions</p>
<h3 id="83-handling-exceptions"><strong>8.3 Handling Exceptions</strong></h3>
<p>You can write programs that handle errors that instead of letting your program break. I assume it uses less memory/processing power than an <code>if</code> statement.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">while</span> <span style="color:#66d9ef">True</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>         x <span style="color:#f92672">=</span> int(input(<span style="color:#e6db74">&#34;Please enter a number: &#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>         <span style="color:#66d9ef">break</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>         print(<span style="color:#e6db74">&#34;Oops!  That was no valid number.  Try again...&#34;</span>)
</span></span></code></pre></div><p>The above code will not break if you were to enter a character or anything other than a number.</p>
<p>Here is how the <code>try</code> statement works.</p>
<ol>
<li>The <em>try clause</em> (between <code>try</code> and <code>except</code>) is executed</li>
<li>If no exception occurs, the <em>except clause</em> is skipped and execution of the <code>try</code> statement is finished</li>
<li>If an exception occurs, the rest of the <em>try clause</em> is skipped. If the type of error matches the exception used in the <code>except</code>, the <em>except clause</em> is executed and then execution continues after the <code>try</code>/<code>except</code> block. (In the example above that doesn&rsquo;t happen, because we put the <code>try</code> statement in a <code>while</code> loop.)</li>
</ol>
<p>A <code>try</code> statement can have more than one <em>except clause</em>, passed in a parenthesized tuple.</p>
<p>You can place a variable after the <code>except</code> keyword, which bounds the exceptions arguments to it.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">Exception</span>(<span style="color:#e6db74">&#39;spam&#39;</span>, <span style="color:#e6db74">&#39;eggs&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> inst:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(type(inst))
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(inst<span style="color:#f92672">.</span>args)     
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(inst)             
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     x, y <span style="color:#f92672">=</span> inst<span style="color:#f92672">.</span>args     
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(<span style="color:#e6db74">&#39;x =&#39;</span>, x)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(<span style="color:#e6db74">&#39;y =&#39;</span>, y)
</span></span><span style="display:flex;"><span><span style="color:#f92672">&lt;</span><span style="color:#66d9ef">class</span> <span style="color:#960050;background-color:#1e0010">&#39;</span><span style="color:#a6e22e">Exception</span><span style="color:#e6db74">&#39;&gt;</span>
</span></span><span style="display:flex;"><span>(<span style="color:#e6db74">&#39;spam&#39;</span>, <span style="color:#e6db74">&#39;eggs&#39;</span>)
</span></span><span style="display:flex;"><span>(<span style="color:#e6db74">&#39;spam&#39;</span>, <span style="color:#e6db74">&#39;eggs&#39;</span>)
</span></span><span style="display:flex;"><span>x <span style="color:#f92672">=</span> spam
</span></span><span style="display:flex;"><span>y <span style="color:#f92672">=</span> eggs
</span></span></code></pre></div><p>We can use <code>Exception</code> as a way to catch most errors if we don&rsquo;t specify the type.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">import</span> sys
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>	f <span style="color:#f92672">=</span> open(<span style="color:#e6db74">&#39;myfile.txt&#39;</span>)
</span></span><span style="display:flex;"><span>	s <span style="color:#f92672">=</span> f<span style="color:#f92672">.</span>readline()
</span></span><span style="display:flex;"><span>	i <span style="color:#f92672">=</span> int(s<span style="color:#f92672">.</span>strip())
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span> <span style="color:#66d9ef">as</span> err:
</span></span><span style="display:flex;"><span>	print(<span style="color:#e6db74">&#34;OS error:&#34;</span>, err)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ValueError</span>:
</span></span><span style="display:flex;"><span>	print(<span style="color:#e6db74">&#34;Could not convert data to an integer.&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> err:
</span></span><span style="display:flex;"><span>	print(<span style="color:#e6db74">f</span><span style="color:#e6db74">&#34;Unexpected </span><span style="color:#e6db74">{</span>err<span style="color:#e6db74">=}</span><span style="color:#e6db74">, </span><span style="color:#e6db74">{</span>type(err)<span style="color:#e6db74">=}</span><span style="color:#e6db74">&#34;</span>)
</span></span><span style="display:flex;"><span>	<span style="color:#66d9ef">raise</span>
</span></span></code></pre></div><p><code>try</code> also has an <em>else clause</em> it can take advantage of. It must follow all <em>except clauses</em>, and is useful for code that must be executed if the try clause does not raise any exception.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">for</span> arg <span style="color:#f92672">in</span> sys<span style="color:#f92672">.</span>argv[<span style="color:#ae81ff">1</span>:]:
</span></span><span style="display:flex;"><span>	<span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span>		f <span style="color:#f92672">=</span> open(arg, <span style="color:#e6db74">&#39;r&#39;</span>)
</span></span><span style="display:flex;"><span>	<span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span>:
</span></span><span style="display:flex;"><span>		print(<span style="color:#e6db74">&#39;cannot open&#39;</span>, arg)
</span></span><span style="display:flex;"><span>	<span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span>		print(arg, <span style="color:#e6db74">&#39;has&#39;</span>, len(f<span style="color:#f92672">.</span>readlines()), <span style="color:#e6db74">&#39;lines&#39;</span>)
</span></span><span style="display:flex;"><span>		f<span style="color:#f92672">.</span>close()
</span></span></code></pre></div><p>Exception handlers also catch errors that happen in the body of code written, not just in the <em>try clause</em>.</p>
<h3 id="84-raising-exceptions"><strong>8.4 Raising Exceptions</strong></h3>
<p><code>raise</code> allows the programmer to force a specified exception to occur.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">NameError</span>(<span style="color:#e6db74">&#39;HiThere&#39;</span>)
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>	File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span>NameErrror: HiThere
</span></span></code></pre></div><p>The following example is fun to play with:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">NameError</span>(<span style="color:#e6db74">&#39;HiThere&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">NameError</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(<span style="color:#e6db74">&#39;An exception flew by!&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span>An exception flew by<span style="color:#960050;background-color:#1e0010">!</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">NameError</span>: HiThere
</span></span></code></pre></div><p>So if I remove the raise on line 5, the code runs with no errors. If you keep it, the code finishes with one exit error. If I don&rsquo;t raise any error in the try, the except obviously doesn&rsquo;t run.</p>
<h3 id="85-exception-chaining"><strong>8.5 Exception Chaining</strong></h3>
<p>This part is a bit complicated, so strap in.</p>
<p>So, there can be instances where an unhandled exception occurs in an except section. In this case, it will attach the error to the raised error.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     open(<span style="color:#e6db74">&#34;database.sqlite&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">OSError</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">RuntimeError</span>(<span style="color:#e6db74">&#34;unable to handle error&#34;</span>)
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">FileNotFoundError</span>: [Errno <span style="color:#ae81ff">2</span>] No such file <span style="color:#f92672">or</span> directory: <span style="color:#e6db74">&#39;database.sqlite&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>During handling of the above exception, another exception occurred:
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">4</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">RuntimeError</span>: unable to handle error
</span></span></code></pre></div><p>You can use the <code>from</code> clause after the <code>raise</code> statement to indicate that an exception is a direct consequence of another.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">func</span>():
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>      <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">ConnectionError</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     func()
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> ConncetionError <span style="color:#66d9ef">as</span> exc:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">RuntimeError</span>(<span style="color:#e6db74">&#39;Failed to open database&#39;</span>) <span style="color:#f92672">from</span> exc
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span>, <span style="color:#f92672">in</span> func
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">ConnectionError</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>The above exception was the direct cause of the following exception:
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">4</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>nodule<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">RuntimeError</span>: Failed to open database
</span></span></code></pre></div><h3 id="86-user-defined-exceptions"><strong>8.6 User-defined Exceptions</strong></h3>
<p>Programs can name their own exceptions by creating a new exception class. They are either directly or indirectly derived from the <code>Exception</code> class.</p>
<p>Common for modules to define their own exceptions to report errors that occur in the functions they define.</p>
<h3 id="87-defining-clean-up-actions"><strong>8.7 Defining Clean-up Actions</strong></h3>
<p>In order to make something execute regardless of exceptions, what we can call &ldquo;clean-up actions&rdquo;, we can use <code>finally</code>.</p>
<p>The way <code>finally</code> interacts with the rest of the <code>try</code> statement can be defined as follows:</p>
<ol>
<li>If an exception occurs during the execution of the <em>try clause</em>, the <code>except</code> statement kicks in, then <code>finally</code> right after. If the error is not handled by the <em>expect clause</em>, the exception is raised after the <code>finally</code> has executed (same thing happens if an exception happens during the <em>exception clause</em></li>
<li>If the <code>finally</code> has a <code>break</code>, <code>continue</code> or <code>return</code> statement, the exceptions are <strong>not</strong> re-raised</li>
<li>If the <code>try</code> statement has a <code>break</code>, <code>continue</code> or <code>return</code> statement, the <code>finally</code> executes just prior to the <code>break</code>, <code>continue</code> or <code>return</code> execution</li>
<li>If the <code>finally</code> has a <code>return</code> statement, the returned value will be the one from the <code>finally</code> instead of the <code>try</code>.</li>
</ol>
<p>Some Examples;</p>
<p>Basic:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>    <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">KeyboardInterrupt</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">finally</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>    print(<span style="color:#e6db74">&#39;Goodbye World!&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span>Goodbye World<span style="color:#960050;background-color:#1e0010">!</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">KeyboardInterrupt</span>
</span></span></code></pre></div><p>Medium:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">bool_return</span>():
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          <span style="color:#66d9ef">return</span> <span style="color:#66d9ef">True</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">finally</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          <span style="color:#66d9ef">return</span> <span style="color:#66d9ef">False</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> bool_return()
</span></span><span style="display:flex;"><span><span style="color:#66d9ef">False</span>
</span></span></code></pre></div><p>Complicated:</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">divide</span>(x, y):
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          result <span style="color:#f92672">=</span> x <span style="color:#f92672">/</span> y
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">ZeroDivisionError</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          print(<span style="color:#e6db74">&#34;division by zero!&#34;</span>)<span style="color:#e6db74">&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">else</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          print(<span style="color:#e6db74">&#34;result is&#34;</span>, result)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">finally</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>          print(<span style="color:#e6db74">&#34;executing finally clause&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> divide(<span style="color:#ae81ff">2</span>, <span style="color:#ae81ff">1</span>)
</span></span><span style="display:flex;"><span>result <span style="color:#f92672">is</span> <span style="color:#ae81ff">2.0</span>
</span></span><span style="display:flex;"><span>executing <span style="color:#66d9ef">finally</span> clause
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> divide(<span style="color:#ae81ff">2</span>, <span style="color:#ae81ff">0</span>)
</span></span><span style="display:flex;"><span>division by zero<span style="color:#960050;background-color:#1e0010">!</span>
</span></span><span style="display:flex;"><span>executing <span style="color:#66d9ef">finally</span> clause
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> divide (<span style="color:#e6db74">&#34;2&#34;</span>, <span style="color:#e6db74">&#34;1&#34;</span>)
</span></span><span style="display:flex;"><span>executing <span style="color:#66d9ef">finally</span> clause
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">3</span>, <span style="color:#f92672">in</span> divide
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">TypeError</span>: unsupprted operand type(s) <span style="color:#66d9ef">for</span> <span style="color:#f92672">/</span>: <span style="color:#e6db74">&#39;str&#39;</span> <span style="color:#f92672">and</span> <span style="color:#e6db74">&#39;str&#39;</span>
</span></span></code></pre></div><h3 id="88-predefined-clean-up-actions"><strong>8.8 Predefined Clean-up Actions</strong></h3>
<p>The only thing this section touches on is that you shouldn&rsquo;t just open a file since it might not close; you should open a file using a <code>with</code> statement.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#66d9ef">with</span> open(<span style="color:#e6db74">&#34;myfile.txt&#34;</span>) <span style="color:#66d9ef">as</span> f:
</span></span><span style="display:flex;"><span>	<span style="color:#66d9ef">for</span> line <span style="color:#f92672">in</span> f:
</span></span><span style="display:flex;"><span>		print(line, end<span style="color:#f92672">=</span><span style="color:#e6db74">&#34;&#34;</span>)
</span></span></code></pre></div><h3 id="89-raising-and-handling-multiple-unrelated-exceptions"><strong>8.9 Raising and Handling Multiple Unrelated Exceptions</strong></h3>
<p>There are situations where there are several exceptions that have occurred. This is the case in concurrency frameworks, when several tasks might&rsquo;ve failed in parallel. <code>ExceptionGroup</code> can list together exception instances.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">def</span> <span style="color:#a6e22e">f</span>():
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     excs <span style="color:#f92672">=</span> [<span style="color:#a6e22e">OSError</span>(<span style="color:#e6db74">&#39;error 1&#39;</span>), <span style="color:#a6e22e">SystemError</span>(<span style="color:#e6db74">&#39;error 2&#39;</span>)]
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> ExceptionGroup(<span style="color:#e6db74">&#39;there were problems&#39;</span>, excs)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> f()
</span></span><span style="display:flex;"><span>  <span style="color:#f92672">+</span> <span style="color:#a6e22e">Exception</span> Group Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  <span style="color:#f92672">|</span>   File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">1</span>, <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span>  <span style="color:#f92672">|</span>   File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">3</span>, <span style="color:#f92672">in</span> f
</span></span><span style="display:flex;"><span>  <span style="color:#f92672">|</span> ExceptionGroup: there were problems
</span></span><span style="display:flex;"><span>  <span style="color:#f92672">+-+----------------</span> <span style="color:#ae81ff">1</span> <span style="color:#f92672">----------------</span>
</span></span><span style="display:flex;"><span>    <span style="color:#f92672">|</span> <span style="color:#a6e22e">OSError</span>: error <span style="color:#ae81ff">1</span>
</span></span><span style="display:flex;"><span>    <span style="color:#f92672">+----------------</span> <span style="color:#ae81ff">2</span> <span style="color:#f92672">----------------</span>
</span></span><span style="display:flex;"><span>    <span style="color:#f92672">|</span> <span style="color:#a6e22e">SystemError</span>: error <span style="color:#ae81ff">2</span>
</span></span><span style="display:flex;"><span>    <span style="color:#f92672">+------------------------------------</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     f()
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     print(<span style="color:#e6db74">f</span><span style="color:#e6db74">&#39;caught</span><span style="color:#e6db74">{</span>type(e)<span style="color:#e6db74">}</span><span style="color:#e6db74">: e&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span>caught <span style="color:#f92672">&lt;</span><span style="color:#66d9ef">class</span> <span style="color:#960050;background-color:#1e0010">&#39;</span><span style="color:#a6e22e">ExceptionGroup</span><span style="color:#e6db74">&#39;&gt;: e</span>
</span></span></code></pre></div><h3 id="810-enriching-exceptions-with-notes"><strong>8.10 Enriching Exceptions with Notes</strong></h3>
<p>If an exception is created in order to be raised, it is usually initialized with information that describes the error that has occurred. Adding information on the error is achievable using the <code>add_note(note)</code> method built-into the exception class.</p>
<div class="highlight"><pre tabindex="0" style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#f92672">&gt;&gt;&gt;</span> <span style="color:#66d9ef">try</span>:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span> <span style="color:#a6e22e">TypeError</span>(<span style="color:#e6db74">&#39;bad type&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span> <span style="color:#66d9ef">except</span> <span style="color:#a6e22e">Exception</span> <span style="color:#66d9ef">as</span> e:
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     e<span style="color:#f92672">.</span>add_note(<span style="color:#e6db74">&#39;Something biffed it dude!&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>     <span style="color:#66d9ef">raise</span>
</span></span><span style="display:flex;"><span><span style="color:#f92672">...</span>
</span></span><span style="display:flex;"><span>Traceback (most recent call last):
</span></span><span style="display:flex;"><span>  File <span style="color:#e6db74">&#34;&lt;stdin&gt;&#34;</span>, line <span style="color:#ae81ff">2</span> <span style="color:#f92672">in</span> <span style="color:#f92672">&lt;</span>module<span style="color:#f92672">&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#a6e22e">TypeError</span>: bad type 
</span></span><span style="display:flex;"><span>Something biffed it dude<span style="color:#960050;background-color:#1e0010">!</span>
</span></span></code></pre></div><hr>
<p>Next:
<a href="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-9/">Chapter 9</a></p>

      </div>
    </article>

    <hr />

    <div class="post-info">
      
    <p>
        <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-tag meta-icon"><path d="M20.59 13.41l-7.17 7.17a2 2 0 0 1-2.83 0L2 12V2h10l8.59 8.59a2 2 0 0 1 0 2.82z"></path><line x1="7" y1="7" x2="7" y2="7"></line></svg>

        <span class="tag"><a href="https://mrpointing.com/tags/computerscience/">computerscience</a></span>
        <span class="tag"><a href="https://mrpointing.com/tags/python/">python</a></span>
        <span class="tag"><a href="https://mrpointing.com/tags/documentation/">documentation</a></span>
        <span class="tag"><a href="https://mrpointing.com/tags/errors/">errors</a></span>
        <span class="tag"><a href="https://mrpointing.com/tags/exceptions/">exceptions</a></span>
        
    </p>

      

      <p>
        <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-file-text">
          <path d="M14 2H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V8z"></path>
          <polyline points="14 2 14 8 20 8"></polyline>
          <line x1="16" y1="13" x2="8" y2="13"></line>
          <line x1="16" y1="17" x2="8" y2="17"></line>
          <polyline points="10 9 9 9 8 9"></polyline>
        </svg>
        1395 Words
      </p>

      <p>
        <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-calendar">
          <rect x="3" y="4" width="18" height="18" rx="2" ry="2"></rect>
          <line x1="16" y1="2" x2="16" y2="6"></line>
          <line x1="8" y1="2" x2="8" y2="6"></line>
          <line x1="3" y1="10" x2="21" y2="10"></line>
        </svg>
        
          2024-06-21 20:00
        

         
          
        
      </p>
    </div>

    
    <div class="pagination">
        

        <div class="pagination__buttons">
            
            <span class="button previous">
                <a href="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-9/">
                    <span class="button__icon">←</span>
                    <span class="button__text">The Python Tutorial Notes - Chapter 9</span>
                </a>
            </span>
            

            
            <span class="button next">
                <a href="https://mrpointing.com/posts/computer-science/python-tutorial/the-python-tutorial-notes---chapter-7/">
                    <span class="button__text">The Python Tutorial Notes - Chapter 7</span>
                    <span class="button__icon">→</span>
                </a>
            </span>
            
        </div>
    </div>


    

    

    

  </main>

            </div>

            
                <footer class="footer">
    
    
</footer>

            
        </div>

        



<script type="text/javascript" src="/bundle.min.e89fda0f29b95d33f6f4224dd9e5cf69d84aff3818be2b0d73e731689cc374261b016d17d46f8381962fb4a1577ba3017b1f23509d894f6e66431f988c00889e.js" integrity="sha512-6J/aDym5XTP29CJN2eXPadhK/zgYvisNc&#43;cxaJzDdCYbAW0X1G&#43;DgZYvtKFXe6MBex8jUJ2JT25mQx&#43;YjACIng=="></script>




    </body>
</html>
